api/osb/reference_instance_plugin.go

```diff
@@ -0,0 +1,368 @@
+
+// NewCheckPlatformIDPlugin creates new plugin that checks the platform_id of the instance
+func NewReferenceInstancePlugin(repository storage.TransactionalRepository, tenantIdentifier string) *referenceInstancePlugin {
+	return &referenceInstancePlugin{
+		repository:       repository,
+		tenantIdentifier: tenantIdentifier,
+	}
+}
+
+// Name returns the name of the plugin
+func (referencePlugin *referenceInstancePlugin) Name() string {
+	return ReferenceInstancePluginName
+}
+
+func (referencePlugin *referenceInstancePlugin) Provision(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	servicePlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	isReferencePlan, err := referencePlugin.isReferencePlan(ctx, "catalog_id", servicePlanID)
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	// Ownership validation
+	path := fmt.Sprintf("context.%s", referencePlugin.tenantIdentifier)
+	callerTenantID := gjson.GetBytes(req.Body, path).String()
+	if callerTenantID != "" {
+		err = referencePlugin.validateOwnership(req)
+		if err != nil {
+			return nil, err
+		}
+	}
+	parameters := gjson.GetBytes(req.Body, "parameters").Map()
+	referencedInstanceID, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+	if !exists {
+		return nil, util.HandleInstanceSharingError(util.ErrMissingReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+	_, err = referencePlugin.isReferencedShared(ctx, referencedInstanceID.String())
+	if err != nil {
+		return nil, err
+	}
+	return referencePlugin.generateOSBResponse(ctx, Provision, nil)
+}
+
+// Deprovision intercepts deprovision requests and check if the instance is in the platform from where the request comes
+func (referencePlugin *referenceInstancePlugin) Deprovision(req *web.Request, next web.Handler) (*web.Response, error) {
+	instanceID := req.PathParams["instance_id"]
+	ctx := req.Context()
+
+	dbInstanceObject, err := referencePlugin.getObjectByOperator(ctx, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+
+	isReferencePlan, err := referencePlugin.isReferencePlan(ctx, "id", instance.ServicePlanID)
+
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	return referencePlugin.generateOSBResponse(ctx, Deprovision, nil)
+}
+
+// UpdateService intercepts update service instance requests and check if the instance is in the platform from where the request comes
+func (referencePlugin *referenceInstancePlugin) UpdateService(req *web.Request, next web.Handler) (*web.Response, error) {
+	// we don't want to allow plan_id and/or parameters changes on a reference service instance
+	instanceID := req.PathParams["instance_id"]
+	ctx := req.Context()
+
+	dbInstanceObject, err := referencePlugin.getObjectByOperator(ctx, types.ServiceInstanceType, "id", instanceID)
```

Comment:
i already saw a getObjectById function used in the filter, do we need another one? if so, what does "ByOperator" mean? the operator is always eq, so i guess this should be `getObjectByField`

Reply:
Moved to the pkg/instance_sharing/utils.go file, with reuse on other places.